### PR TITLE
Improved conditional PDF/CDF/Quantile

### PR DESCRIPTION
**Pull request by regislebrun**
The main contribution of this PR is the introduction to sequential evaluations of conditional PDF/CDF/Quantile in order to speed-up the Rosenblatt transformation. The gain ranges from significant (+30%) to huge (x100) depending on the distribution.
Other contributions are:
*  A bug fix in Dirichlet conditional services, where the sum of conditioning components was confused with the sum of marginal theta parameters
* A specific high precision implementation of Spearman correlation for the AliMikhailHaq copula
* A vectorized implementation of pNormal, qNormal, pStudent and qStudent in DistFunc.

The following benchmark allows to quantify the improvement:
```
from openturns import *
from time import time

# Some helper functions
def conditionalPDF(distribution, inS):
    try:
        res = Sample([distribution.computeSequentialConditionalPDF(x) for x in inS])
    except:
        size = inS.getSize()
        dim = distribution.getDimension()
        res = Sample(size, dim)
        for i in range(size):
            y = Point(0)
            for j in range(dim):
                x = inS[i, j]
                res[i, j] = distribution.computeConditionalPDF(x, y)
                y.add(x)
    return res
        
def conditionalCDF(distribution, inS):
    try:
        res = Sample([distribution.computeSequentialConditionalCDF(x) for x in inS])
    except:
        size = inS.getSize()
        dim = distribution.getDimension()
        res = Sample(size, dim)
        for i in range(size):
            y = Point(0)
            for j in range(dim):
                x = inS[i, j]
                res[i, j] = distribution.computeConditionalCDF(x, y)
                y.add(x)
    return res
        
def conditionalQuantile(distribution, inS):
    try:
        if not new:
            raise
        res = Sample([distribution.computeConditionalQuantile(q) for q in inS])
    except:
        size = inS.getSize()
        dim = distribution.getDimension()
        res = Sample(size, dim)
        for i in range(size):
            y = Point(0)
            for j in range(dim):
                q = inS[i, j]
                res[i, j] = distribution.computeConditionalQuantile(q, y)
                y.add(res[i, j])
    return res

# The benchmark
def bench(distribution):
    RandomGenerator.SetSeed(0)
    print("=+="*30)
    print("distribution=", distribution.getClassName())
    inS = distribution.getSample(N)
    print("inS=")
    print(inS)
    
    print("#"*10)
    try:
        t0 = time()
        res = conditionalPDF(distribution, inS)
        t1 = time()
        print("Speed conditional PDF=", N / (t1 - t0), "evals/s")
        print(res)
    except:
        print("Skip conditional PDF")

    print("#"*10)
    try:
        t0 = time()
        res = conditionalCDF(distribution, inS)
        t1 = time()
        print("Speed conditional CDF=", N / (t1 - t0), "evals/s")
        print(res)
    except:
        print("Skip conditional CDF")
        
    print("#"*10)
    try:
        t0 = time()
        res = conditionalQuantile(distribution, res)
        t1 = time()
        print("Speed conditional quantile=", N / (t1 - t0), "evals/s")
        print(res)
    except:
        print("Skip conditional quantile")
        
    print("#"*10)
    try:
        iso = distribution.getIsoProbabilisticTransformation()
        t0 = time()
        res = iso(inS)
        t1 = time()
        print("Speed iso=", N / (t1 - t0), "evals/s")
        print(res)
    except:
        print("Skip isoprobabilistic transformation")

    print("#"*10)
    try:
        invIso = distribution.getInverseIsoProbabilisticTransformation()
        t0 = time()
        res = invIso(res)
        t1 = time()
        print("Speed invIso=", N / (t1 - t0), "evals/s")
        print(res)
    except:
        print("Skip inverse isoprobabilistic transformation")

    print("#"*10)

# The test-cases

N = 10

dim = 5
size = 100
sample = Normal(dim).getSample(size)

C = CovarianceMatrix(dim)
for i in range(dim):
    C[i,i] = dim**2
    for j in range(i):
        C[i,j] = ((i + j)/2)**2

from math import sqrt

dim = C.getDimension()
sigma = [sqrt(C[i, i]) for i in range(dim)]
R = CorrelationMatrix(dim)
for i in range(dim):
    for j in range(i):
        R[i, j] = C[i,j] / (sigma[i] * sigma[j])

nu = 5.0

for d in [KernelMixture(Normal(), Point(dim, 1.0), sample), Normal([i for i in range(dim)], C), Mixture([ComposedDistribution([Gumbel() + k for k in x]) for x in sample]), Mixture([Normal(x, C) for x in sample]), Dirichlet([1.0]*dim), Student(nu, [i for i in range(dim)], sigma, R), ComposedCopula([OrdinalSumCopula([ClaytonCopula(0.5), GumbelCopula(1.5)], [0.5]), NormalCopula(CorrelationMatrix(2, [1.0, 0.5, 0.5, 1.0]))])]:
    bench(d)
```
The result I got using either master or this branch are given below:
[res_branch.txt](https://github.com/openturns/openturns/files/3036101/res_branch.txt)
[res_master.txt](https://github.com/openturns/openturns/files/3036102/res_master.txt)

### CODE REVIEW

**Comment by jschueller** [comment]
Some tests produce way to much digits, I suggest "%.12g" formatting, tops
```
-conditional CDF= 0.9020865618683835
+conditional CDF= 0.9020865618683833
```

File: lib/src/Uncertainty/Distribution/DistFunc.cxx [hunk 1]
```diff
@@ -978,6 +992,97 @@ Scalar DistFunc::qNormal(const Scalar p,
...
 
+Point DistFunc::qNormal(const Point & p,
+                         const Bool tail)
+{
+  const UnsignedInteger size = p.getSize();
+  Point result(size);
+  for (UnsignedInteger i = 0; i < size; ++i)
+    result[i] = qNormal(p[i], tail);
+  return result;
+  #ifdef toto
```
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/DistFunc.cxx]: should we clean this up ?

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx [hunk 2]
```diff
@@ -85,7 +85,7 @@ TestResult LinearModelTest::LinearModelFisher(const Sample & firstSample,
 TestResult LinearModelTest::LinearModelFisher(const Sample & firstSample,
     const Sample & secondSample,
     const LinearModel & linearModel,
-    const Scalar level)
+    const Scalar )
```
Comment (jschueller) [inline lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: level was probably missing in the call to `LinearModelFisher(firstSample, secondSample, linearModelResult)`

**Comment by jschueller** [comment]
Could you rebase after #1063 changes ?

File: lib/src/Uncertainty/Distribution/DistFunc.cxx (hunk 1 above)
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx]: done

File: lib/src/Uncertainty/StatTests/LinearModelTest.cxx (hunk 2 above)
Comment (regislebrun) [reply lib/src/Uncertainty/StatTests/LinearModelTest.cxx]: done

File: lib/src/Uncertainty/Distribution/KernelMixture.cxx [hunk 3]
```diff
@@ -489,6 +489,135 @@ Point KernelMixture::computeCDFGradient(const Point & point) const
...
+  Scalar jointCDF = 0.0;
+  Scalar marginalPDF = 0.0;
+  Scalar h = bandwidth_[conditioningDimension];
+  for (UnsignedInteger i = 0; i < size; ++i)
+  {
+    Scalar marginalAtomPDF = kernel_.computePDF((y[0] - sample_(i, 0)) / bandwidth_[0]);
+    for (UnsignedInteger j = 1; j < conditioningDimension; ++j)
+      marginalAtomPDF *= kernel_.computePDF((y[j] - sample_(i, j)) / bandwidth_[j]);
+    marginalPDF += marginalAtomPDF;
+    jointCDF += marginalAtomPDF * kernel_.computeCDF((x - sample_(i, conditioningDimension)) / h);
```
Comment (sofianehaddad) [inline lib/src/Uncertainty/Distribution/KernelMixture.cxx]: hmm when integrating I think the bandwidth_[j]) should vanish. Isn't it ?

**Pull request merged**